### PR TITLE
Allow nonce to be overridden

### PR DESCRIPTION
**Pull request by carl**
The nonce is supposed to be persisted (or replicable) locally in order to verify it on OAuth completion. That means it needs to be possible to pass it in, rather than have it generated internally.

See https://help.shopify.com/api/guides/authentication/oauth:

> {nonce} - a randomly selected value provided by your application, which is unique for each authorization request. During the OAuth callback phase, your application must check that this value matches the one you provided during authorization. This mechanism is important for the security of your application.

### CODE REVIEW

**Comment by lpinca** [comment]
You can it from the auth URL, see https://github.com/lpinca/shopify-token/blob/master/example/index.js#L32.

**Comment by carl** [comment]
Yes we could get it from the auth url from the internally generated nonce, however that would require persisting it somewhere in order to validate the request later. The other way to achieve this is to simply hash some known value like the shop name with a secret, which is why we would like to add this optional functionality.

**Comment by benbuckman** [comment]
@lpinca – See the docs,

> During the OAuth callback phase, your application must check that this value matches the one you provided during authorization.

If you just take it from the final URL, how do you verify that it's correct?

**Comment by lpinca** [comment]
@benbuckman you get it from the URL and save it somewhere see the above example.
The docs say that a new **unique** `nonce` must be generated for each authorization request so my point is... Instead of doing:

```js
var nonce = shopifyToken.generateNonce();
var url = shopifyToken.generateAuthUrl('foo', undefined, nonce);
```

just grab the nonce from the generate URL:

```js
var uri = shopifyToken.generateAuthUrl('foo');
var nonce = url.parse(uri, true).query.state;
```

This feature will make sense if you want to use a custom nonce but I can't see any compelling reason to do that. If you have a particular use case, please share.

**Comment by benbuckman** [comment]
```js
var nonce = url.parse(uri, true).query.state;
```

is very clunky. You need the nonce outside the method where it's created, but because it's created inside that method, have to jump through hoops to get it outside that method... You're also only allowing the nonce to be generated one way (`crypto.randomBytes`) when there are many valid ways it can be generated.

**Comment by lpinca** [comment]
@benbuckman you are right it's not the best way, I'll add some comments to this pr, it needs some work.

File: index.js [hunk 1]
```diff
@@ -59,20 +59,26 @@ function ShopifyToken(options) {
   this.apiKey = options.apiKey;
 }
 
+ShopifyToken.prototype.generateNonce = function() {
```
Comment (lpinca) [inline index.js]: Please add JSDoc for this method and also add documentation for it in the README.

File: index.js [hunk 2]
```diff
@@ -59,20 +59,26 @@ function ShopifyToken(options) {
...
  * @param {String} shop The shop name
  * @param {Array|String} [scopes] The list of scopes
+ * @param {String} Random string (ideally persisted elsewhere) for comparison on callback
  * @return {String} The authorization URL
  * @public
  */
-ShopifyToken.prototype.generateAuthUrl = function generateAuthUrl(shop, scopes) {
-  scopes || (scopes = this.scopes);
+ShopifyToken.prototype.generateAuthUrl = function generateAuthUrl(shop, scopes, nonce) {
+  scopes = scopes || this.scopes;
```
Comment (lpinca) [inline index.js]: Allow both `scopes` and `nonce` to be optional. Right now if `scopes` is not provided but `nonce` is, `scopes` will get the `nonce` value.

File: index.js [hunk 3]
```diff
@@ -59,20 +59,26 @@ function ShopifyToken(options) {
...
+ShopifyToken.prototype.generateNonce = function() {
+  return crypto.randomBytes(16).toString('hex');
+};
+
 /**
  * Build the authorization URL.
  *
  * @param {String} shop The shop name
  * @param {Array|String} [scopes] The list of scopes
+ * @param {String} Random string (ideally persisted elsewhere) for comparison on callback
```
Comment (lpinca) [inline index.js]: Add the argument name (`nonce`) and please try to respect the 80 chars limit.

File: index.js [hunk 4]
```diff
@@ -59,20 +59,26 @@ function ShopifyToken(options) {
...
  * @return {String} The authorization URL
  * @public
  */
-ShopifyToken.prototype.generateAuthUrl = function generateAuthUrl(shop, scopes) {
-  scopes || (scopes = this.scopes);
+ShopifyToken.prototype.generateAuthUrl = function generateAuthUrl(shop, scopes, nonce) {
```
Comment (lpinca) [inline index.js]: Please also update the documentation in the README and specify that if `nonce` is not provided it will be generate automatically.

File: test.js [hunk 5]
```diff
@@ -55,30 +55,32 @@ describe('shopify-token', function () {
   describe('#generateAuthUrl', function () {
     it('builds the authorization URL', function () {
       var uri = shopifyToken.generateAuthUrl('qux');
+      var nonce = url.parse(uri, true).query.state;
```
Comment (lpinca) [inline test.js]: Use `shopifyToken.generateNonce`.

**Comment by lpinca** [comment]
Please also add a test for `ShopifyToken.prototype.generateNonce()`.

File: index.js [hunk 6]
```diff
@@ -59,20 +59,26 @@ function ShopifyToken(options) {
...
  * Build the authorization URL.
  *
  * @param {String} shop The shop name
  * @param {Array|String} [scopes] The list of scopes
+ * @param {String} Random string (ideally persisted elsewhere) for comparison on callback
  * @return {String} The authorization URL
  * @public
  */
-ShopifyToken.prototype.generateAuthUrl = function generateAuthUrl(shop, scopes) {
-  scopes || (scopes = this.scopes);
```
Comment (lpinca) [inline index.js]: Unnecessary change, keep it.

File: index.js [hunk 7]
```diff
@@ -59,20 +59,26 @@ function ShopifyToken(options) {
...
+ShopifyToken.prototype.generateAuthUrl = function generateAuthUrl(shop, scopes, nonce) {
+  scopes = scopes || this.scopes;
+  nonce = nonce || this.generateNonce();
```
Comment (lpinca) [inline index.js]: Put this below, in query object so we don't reassign the argument.

**Comment by lpinca** [comment]
Merged as b517fb4deed64f1de795cc922461771ac44606d9, thanks!

**Comment by lpinca** [comment]
I'll cut a new major version tomorrow as the changes to the error messages in 95317751e6575af9a6e51ca19eb82728327367b0 are breaking changes.
